File: 088.LeetCode@98_Validate_Binary_Search_Tree/088.LeetCode@98_Validate_Binary_Search_Tree.py
class Solution:
    def isValidBST(self, root):
        return self.helper(root, None, None)

    def helper(self, root, min_node, max_node):
        if root is None:
            return True
        if min_node is not None and root.val <= min_node.val:
            return False
        if max_node is not None and root.val >= max_node.val:
            return False
        return self.helper(root.left, min_node, root) and self.helper(root.right, root, max_node)

File: 088.LeetCode@98_Validate_Binary_Search_Tree/088.LeetCode@98_Validate_Binary_Search_Tree.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, null,null);
    }
    public boolean helper(TreeNode root, TreeNode min, TreeNode max){
        if (root==null)return true;
        if (min!=null && root.val<=min.val) return false;
        else if (max!=null && root.val>=max.val) return false;
        return helper(root.left,min,root) && helper(root.right,root,max);
    }
}